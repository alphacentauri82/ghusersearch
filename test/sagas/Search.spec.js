import {
  call,
} from 'redux-saga/effects';
import {isFSA} from 'flux-standard-action';
import api from 'store/api';
import {
  searchUsers,
} from 'store/sagas/Search';

describe('Saga: searchUsers', () => {

  describe('when a cached result exists', () => {
    it('should pass the result to the success action and not call the API', () => {
      const generator = searchUsers({payload: {search: 'q=simon'}});
      generator.next();

      const successAction = generator.next({data: 'foo'}).value;
      expect(successAction.PUT.action).toMatchSnapshot();

      const rateLimitAction = generator.next().value;
      expect(rateLimitAction.PUT.action).toMatchSnapshot();

      expect(isFSA(successAction.PUT.action)).toBeTruthy();
      expect(isFSA(rateLimitAction.PUT.action)).toBeTruthy();
    });
  });

  describe('when no cached data exists', () => {
    it('should call the API and put a success action with the payload', () => {
      const generator = searchUsers({payload: {search: 'q=simon'}});
      generator.next();
      const apiCall = generator.next().value;
      expect(apiCall).toEqual(
        call(api.searchUsers, {q: 'simon'})
      );

      const response = {
        pagination: null,
        total_count: 2,
        items: [
          {id: 123, avatar_url: 'foo', login: 'alecrust', type: 'User'},
          {id: 456, avatar_url: 'foo', login: 'simonsmith', type: 'User'},
        ],
      };
      const successAction = generator.next(response).value;
      expect(successAction.PUT.action).toMatchSnapshot();
    });
  });

  describe('when a request to the API fails', () => {
    it('should call a failure action', () => {
      const generator = searchUsers({payload: {search: 'q=simon'}});
      const apiError = new Error('it went wrong');
      apiError.response = 'test';

      generator.next();
      generator.next();
      const errorAction = generator.throw(apiError).value;
      expect(errorAction.PUT.action).toMatchSnapshot();
    });
  });

});
