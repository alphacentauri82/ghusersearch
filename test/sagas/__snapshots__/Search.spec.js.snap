// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga: searchUsers when a cached result exists should pass the result to the success action and not call the API 1`] = `
Object {
  "meta": Object {
    "fromCache": true,
  },
  "payload": Object {
    "data": "foo",
    "search": "q=simon",
  },
  "type": "SEARCH_SUCCESS",
}
`;

exports[`Saga: searchUsers when a cached result exists should pass the result to the success action and not call the API 2`] = `
Object {
  "meta": Object {
    "fromCache": true,
  },
  "type": "API_RATE_LIMIT_REQUEST",
}
`;

exports[`Saga: searchUsers when a request to the API fails should call a failure action 1`] = `
Object {
  "error": true,
  "payload": [Error: it went wrong],
  "type": "SEARCH_FAILURE",
}
`;

exports[`Saga: searchUsers when no cached data exists should call the API and put a success action with the payload 1`] = `
Object {
  "meta": Object {
    "fromCache": false,
  },
  "payload": Object {
    "entities": Object {
      "users": Object {
        "123": Object {
          "avatar_url": "foo",
          "id": 123,
          "login": "alecrust",
        },
        "456": Object {
          "avatar_url": "foo",
          "id": 456,
          "login": "simonsmith",
        },
      },
    },
    "pagination": null,
    "result": Array [
      123,
      456,
    ],
    "search": "q=simon",
    "totalResults": 2,
  },
  "type": "SEARCH_SUCCESS",
}
`;
